Add TODO form representing incomplete programs

david-christiansen opened the pull request:
A "shed" is a region of a program that is not yet written, in which the programmer can write whatever they want. The intention behind sheds is that they be used for fancy interactive editing features. Haskell has a version of sheds, which they call "typed holes", that don't allow any contents. These ones are more similar to those found in Epigram or Agda.

At present, a shed macroexpands to a program that throws an error that lists the expected type. However, they also add a syntax property that tools such as the todo list can use to provide a list of sheds and their goal types.

Future enhancements can include diffing the type synthesized for the shed contents against the type expected for the shed itself, and adding the typing context to the message and/or goal. Sheds in types would also be nice. But this feature is, I think already useful as it currently is.

lexi-lambda commented:
This is cool. A few comments.

  1. Is the “shed” terminology really common? It seem specific to Epigram, whereas Haskell and Agda both call them “holes”.

  2. This sort of thing seems like it maybe ought to replace `undefined!`. Keeping `error!` around makes some sense, since it’s sometimes used for signaling when “the impossible happened”, but `undefined!` is really just for leaving parts of the program unimplemented.

  3. Whatever the interface is, it should be documented.

Thoughts?

david-christiansen commented:
Perhaps "shed" is indeed too obscure :-)

How about uniting `undefined!` and sheds, calling the form `TODO` or `todo` instead? That seems to communicate the purpose of the form quite clearly, while at the same time not seeming particularly exotic.  At the same time, it might help those newer to working with types figure out the difference between `undefined` and `error`.

In Scala, `???` is the equivalent of `undefined`, which also seems to be a fine name for those who are more punctuation-inclined.

Whatever name you like best, I'm happy to use it.

david-christiansen commented:
I added docs and renamed it to `TODO`. I didn't delete `undefined!`, because `TODO` is a special form rather than a value that works at any type.

What do you think?

lexi-lambda commented:
Sorry for my lack of comments on this. I’ve been very busy with work, and I’d hoped to get some time to look at this over the holiday, but truthfully, I’ve been exhausted, so I ended up spending basically the whole time catching up on sleep and relaxing.

The truth of the matter is that I’m still conflicted about a few things.

  1. I don’t know if this should replace `undefined!` or not. It *seems* like it largely can by making `TODO` also work as an identifier macro that works like `undefined!` currently does.

  2. Honestly, I find the fact that you can put arbitrary code inside it that’s completely ignored a bit… weird? The idea of eventually inferring and subsequently diffing the types makes some sense, but how does that work with global, unification-based type inference? What happens if the code inside the hole is just ill-typed? Currently, it seems like this would be equally well served with a standalone `TODO` followed by a `#;` expression comment.

If I were building this feature myself, I’d probably do two things. I’d change the name to `todo!`, replace `undefined!` with it, and turn it into an identifier macro. When used in function position, i.e. `(todo! ...)`, I would make it so you could specify a string literal inside the parentheses that includes a message that would show up in your DrRacket plugin.

However, this is something with a purpose that you understand better than I do. Maybe there’s some workflow this feature enables that I’m not currently aware of that would motivate it a little better in my mind?

david-christiansen commented:
I hope that you feel under no obligation whatsoever to do free coding/reviewing work for me. Thanks for taking time now that you're recovered a bit!

The idea of a shed/hole is that it's a place for the user to stash things that are not yet ready for the type checker. Eventually, interactive commands can use those stashed things to help write the program. For instance, in `agda-mode` for Emacs, one can write a name in the hole and then issue the "refine" command, which unifies the type of the name that was written with the type of the hole, constructing a suitable number of holes for arguments. It is also possible to attempt to synthesize a type for whatever is in the hole and then see that synthesized type next to the type that the hole must have, which can provide insight. If the types match, then the hole wrapping an expression can be removed. For doing type-driven synthesis, a list of hints of names to use can be put in the hole.

These modes of interaction require that the shed/hole can hold basically arbitrary expressions. If a task name string is what's needed, then I'd want to provide that with a keyword argument to the macro.

You know better than I whether this should replace `undefined!`.

lexi-lambda commented:
> In `agda-mode` for Emacs, one can write a name in the hole and then issue the "refine" command, which unifies the type of the name that was written with the type of the hole, constructing a suitable number of holes for arguments. It is also possible to attempt to synthesize a type for whatever is in the hole and then see that synthesized type next to the type that the hole must have, which can provide insight.

Do you think this is feasible in a language with Haskell-style type inference? Agda has rather different type inference properties than Haskell/Hackett.

david-christiansen commented:
I think it's completely feasible. When showing the difference, the type of an expression in the hole should be such that some subset of its \forall-quantified variables can be instantiated to make it alpha-equivalent to the type expected for the hole (and such that type class constraints have a solution). In other words, it should be a subtype.

Example:
```
(: (TODO show) (-> Integer String))
```
might say:
```
At foo.rkt line 30 column 3:
  The expression
    show
  has type
    (forall (a) (Show a) => (-> a String))
  where the type
    (-> Integer String)
  is expected.  These types are compatible.
```
modulo curly-brace infix and the like. The right-click menu could then support "Merge expression" to delete the surrounding TODO form.

david-christiansen commented:
I think that the version in which the contents of the TODO are delayed until after other type inference, with all the scary non-local effects that you're describing, is exactly the right one for an ML or Haskell-style type system, and would be quite useful in practice.

lexi-lambda commented:
I think that makes sense. That means that type errors would currently prevent holes from producing a result, which is a problem, but that needs to be fixed in general anyway (making it possible for programs to produce multiple type errors).

I’ll look into merging this when I get a free moment.

lexi-lambda commented:
Finally merged. Thanks for this improvement! 😄
